src/core/net/tcp6.cpp

```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
     return error;
 }
 
-Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
+bool Tcp::Listener::Matches(const MessageInfo &aMessageInfo) const
 {
-    OT_UNUSED_VARIABLE(aIp6Header);
-    OT_UNUSED_VARIABLE(aMessage);
-    OT_UNUSED_VARIABLE(aMessageInfo);
+    bool matches = false;
+
+    VerifyOrExit(mTcbListen.t_state == TCP6S_LISTEN);
+    VerifyOrExit(mTcbListen.lport == HostSwap16(aMessageInfo.GetSockPort()));
+    VerifyOrExit(GetLocalIp6Address().IsUnspecified() || GetLocalIp6Address() == aMessageInfo.GetSockAddr());
+
+    matches = true;
+
+exit:
+    return matches;
+}
 
+Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
+{
     Error error = kErrorNotImplemented;
 
-    for (Endpoint &active : mEndpoints)
+    /*
+     * The type uint32_t was chosen for alignment purposes. The size is the
+     * maximum TCP header size, including options.
+     */
+    uint32_t header[15];
+
+    uint16_t length = aIp6Header.GetPayloadLength();
+    uint8_t  headerSize;
+
+    struct ip6_hdr *ip6Header;
+    struct tcphdr * tcpHeader;
+
+    Endpoint *endpoint;
+    Endpoint *endpointPrev;
+
+    Listener *listener;
+    Listener *listenerPrev;
+
+    VerifyOrExit(length == aMessage.GetLength() - aMessage.GetOffset(), error = kErrorParse);
+    VerifyOrExit(length >= sizeof(Tcp::Header), error = kErrorParse);
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), &headerSize, 1));
```

Comment:
Use `sizeof(headersize)` instead of `1`?

Reply:
Actully, instead of 
```c++
Error Message::Read(uint16_t aOffset, void *aBuf, uint16_t aLength) const;
```
you can use the template version of `Message::Read<ObjectType>` to read a `uint8_t` value directly.
```c++
    template <typename ObjectType> Error Read(uint16_t aOffset, ObjectType &aObject) const
```
The code will be simpler (the compiler should infer the `ObjectType` to be `uint8_t` from the type of `headerSize`):
```c++
SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), headerSize));
```